### PR TITLE
Custom email

### PR DESCRIPTION
**Pull request by Miodec**
This pr adds support for sending custom emails.

TODO:
 - [x] Add prometheus logging for amount of emails sent
 - [x] Generate and send verification email from the backend
   - [x] Remove verification sending from frontend 
 - [x] Send password reset emails from the backend
   - [x] Remove this from the frontend, send an API request instead
   - [x] Prepare template

Potentially also in this PR or future PR:
 - [ ] Add template for sending a support code email
 - [ ] Send emails when critical things happen like email and password changes

### CODE REVIEW

File: backend/src/constants/email-templates/verification.html [hunk 1]
```diff
@@ -0,0 +1,84 @@
+<mjml>
```
Comment (Bruception) [inline backend/src/constants/email-templates/verification.html:1]: Constants probably isn;t the best place for this. Create new dir: email-templates

File: backend/src/utils/email.ts [hunk 2]
```diff
@@ -0,0 +1,149 @@
+import * as nodemailer from "nodemailer";
```
Comment (Bruception) [inline backend/src/utils/email.ts:1]: Probably want to move this to `init` instead.
Rename file to `email-client`

File: backend/src/server.ts [hunk 3]
```diff
@@ -43,16 +48,20 @@ async function bootServer(port: number): Promise<Server> {
       const connection = RedisClient.getConnection();
 
       Logger.info("Initializing queues...");
+      const queueNames: string[] = [];
       queues.forEach((queue) => {
         queue.init(connection);
+        queueNames.push(queue.queueName);
       });
-      Logger.success("Queues initialized");
+      Logger.success("Queues initialized: " + queueNames.join(", "));
```
Comment (Bruception) [inline backend/src/server.ts:56]: Logger.success("Queues initialized: " + queues.map(queue => queue.name).join(", "));

File: backend/src/server.ts [hunk 4]
```diff
@@ -35,6 +37,9 @@ async function bootServer(port: number): Promise<Server> {
     const liveConfiguration = await getLiveConfiguration();
     Logger.success("Live configuration fetched");
 
+    Logger.info("Initializing email...");
+    await Email.init();
```
Comment (Bruception) [inline backend/src/server.ts:41]: Nit:
Email -> EmailClient

```
Logger.info("Initializing email client...");
```

File: backend/src/utils/email.ts [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+      Logger.warning("No email configuration provided. Running without email.");
+      return;
+    }
+    throw new Error("No email configuration provided");
+  }
+
+  transporter = nodemailer.createTransport({
+    host: EMAIL_HOST,
+    secure: EMAIL_PORT === "465" ? true : false,
+    port: parseInt(EMAIL_PORT ?? "578"),
```
Comment (Bruception) [inline backend/src/utils/email.ts:33]:
```
parseInt(..., 10)
```

File: backend/src/utils/email.ts [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+    Logger.info("Verifying email configuration...");
+    const result = await transporter.verify();
+
+    if (result !== true) {
+      throw new Error(
+        `Could not verify email configuration: ` + JSON.stringify(result)
+      );
+    }
+
+    Logger.success("Email configuration verified");
```
Comment (Bruception) [inline backend/src/utils/email.ts:50]: Email -> Email client

File: backend/src/utils/email.ts [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+
+export async function sendMail(
+  to: string,
+  subject: string,
+  html: string
+): Promise<MailResult> {
+  if (!isInitialized()) {
+    return {
+      success: false,
+      message: "Email transport not initialized",
```
Comment (Bruception) [inline backend/src/utils/email.ts:80]: Email client

File: backend/src/utils/email.ts [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+);
+
+const cachedTemplates: Record<string, string> = {};
+
+function getTemplate(name: string): string {
+  if (cachedTemplates[name]) {
+    return cachedTemplates[name];
+  }
+
+  const template = fs.readFileSync(
```
Comment (Bruception) [inline backend/src/utils/email.ts:131]: Use async read file API, this can block the event loop

File: backend/src/utils/email.ts [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+  const html = mjml2html(template).html;
+
+  cachedTemplates[name] = html;
+  return html;
+}
+
+export function fillTemplate(type: EmailType, data: any[]): string {
+  if (type === "verify") {
+    return getTemplate("verification.html")
+      .replace(/{{name}}/gim, data[0])
```
Comment (Bruception) [inline backend/src/utils/email.ts:145]: Probably want to use something like mustache instead

File: backend/package.json [hunk 10]
```diff
@@ -16,6 +16,7 @@
     "npm": "8.1.2"
   },
   "dependencies": {
+    "@types/mustache": "^4.2.2",
```
Comment (Bruception) [inline backend/package.json:19]: Dev dependency

File: backend/package.json [hunk 11]
```diff
@@ -55,6 +59,7 @@
     "@types/lodash": "4.14.178",
     "@types/node": "17.0.18",
     "@types/node-fetch": "2.6.1",
+    "@types/nodemailer": "^6.4.7",
```
Comment (Bruception) [inline backend/package.json:62]: dev dependency

File: backend/src/server.ts [hunk 12]
```diff
@@ -46,13 +51,19 @@ async function bootServer(port: number): Promise<Server> {
       queues.forEach((queue) => {
         queue.init(connection);
       });
-      Logger.success("Queues initialized");
+      Logger.success(
+        "Queues initialized: " +
```
Comment (Bruception) [inline backend/src/server.ts:55]: use string interpolation: 
```ts
`Queues initialized ${...}`
```

File: backend/src/server.ts [hunk 13]
```diff
@@ -35,6 +37,9 @@ async function bootServer(port: number): Promise<Server> {
     const liveConfiguration = await getLiveConfiguration();
     Logger.success("Live configuration fetched");
 
+    Logger.info("Initializing email...");
```
Comment (Bruception) [inline backend/src/server.ts:40]:
```
 Logger.info("Initializing email client...");
 ```

File: backend/src/init/email-client.ts [hunk 14]
```diff
@@ -0,0 +1,177 @@
...
+    secure: EMAIL_PORT === "465" ? true : false,
+    port: parseInt(EMAIL_PORT ?? "578", 10),
+    auth: {
+      user: EMAIL_USER,
+      pass: EMAIL_PASS,
+    },
+  });
+
+  try {
+    Logger.info("Verifying email configuration...");
```
Comment (Bruception) [inline backend/src/init/email-client.ts:42]: email-client

File: backend/src/types/types.d.ts [hunk 15]
```diff
@@ -474,4 +474,24 @@ declare namespace MonkeyTypes {
     canGetPb: boolean;
     difficultyLevel: number;
   }
+
+  type EmailType = "verify" | "resetPassword";
```
Comment (Bruception) [inline backend/src/types/types.d.ts:478]: Retrospectively, putting all types in a namespace is not the best choice.
A better way to do this is to just export types from modules.

Let's keep these types within the EmailQueue.

File: backend/src/workers/index.ts [hunk 16]
```diff
@@ -1,3 +1,4 @@
 import LaterWorker from "./later-worker";
+import Emailer from "./email-worker";
```
Comment (Bruception) [inline backend/src/workers/index.ts:2]: Emailer -> EmailWorker for consistency

File: backend/src/init/email-client.ts [hunk 17]
```diff
@@ -0,0 +1,179 @@
...
+
+const cachedTemplates: Record<string, string> = {};
+
+async function getTemplate(name: string): Promise<string> {
+  if (cachedTemplates[name]) {
+    return cachedTemplates[name];
+  }
+
+  const template = await fs.promises.readFile(
+    EMAIL_TEMPLATES_DIRECTORY + "/" + name,
```
Comment (Bruception) [inline backend/src/init/email-client.ts:160]: Use string interpolation `${EMAIL_TEMPLATES_DIRECTORY}/${name}`

File: backend/src/init/email-client.ts [hunk 18]
```diff
@@ -0,0 +1,179 @@
...
+  const result = await transporter.sendMail(mailOptions);
+
+  recordEmail(templateName, result.accepted.length === 0 ? "fail" : "success");
+
+  if (result.accepted.length === 0) {
+    return {
+      success: false,
+      message: result.response,
+    };
+  } else {
```
Comment (Bruception) [inline backend/src/init/email-client.ts:106]: No need for else

File: backend/src/init/email-client.ts [hunk 19]
```diff
@@ -0,0 +1,137 @@
...
+    subject,
+    html: template,
+  };
+
+  const result = await transporter.sendMail(mailOptions);
+
+  recordEmail(templateName, result.accepted.length === 0 ? "fail" : "success");
+
+  return {
+    success: result.accepted.length !== 0,
```
Comment (Bruception) [inline backend/src/init/email-client.ts:103]: nice

File: backend/src/workers/email-worker.ts [hunk 20]
```diff
@@ -0,0 +1,38 @@
...
+  const type: EmailType = job.data.type;
+  const email: string = job.data.email;
+  const ctx: EmailTaskContexts[typeof type] = job.data.ctx;
+
+  Logger.info(`Starting job: ${type}`);
+
+  const start = performance.now();
+
+  if (type === "verify") {
+    await sendMailUsingTemplate(
```
Comment (Bruception) [inline backend/src/workers/email-worker.ts:21]: So right now, we are not consuming the response from this right?
On failure we will want to mark this as a failed attempt.

File: backend/src/queues/email-queue.ts [hunk 21]
```diff
@@ -0,0 +1,58 @@
+import { MonkeyQueue } from "./monkey-queue";
+
+const QUEUE_NAME = "email-tasks";
+
+export type EmailType = "verify" | "resetPassword";
+
+export interface EmailTask<M extends EmailType> {
```
Comment (Bruception) [inline backend/src/queues/email-queue.ts:7]: niceee

File: backend/src/server.ts [hunk 22]
```diff
@@ -62,6 +75,14 @@ async function bootServer(port: number): Promise<Server> {
...
 
     recordServerVersion(version);
+    //testing email queue
+    if (process.env.EMAIL_USER) {
+      emailQueue.sendVerificationEmail(
+        process.env.EMAIL_USER,
+        "Miodec",
+        "https://monkeytype.com"
+      );
+    }
```
Comment (Bruception) [inline backend/src/server.ts:85]: I'm assuming this will be removed.

File: backend/src/server.ts [hunk 23]
```diff
@@ -14,6 +14,8 @@ import * as RedisClient from "./init/redis";
 import queues from "./queues";
 import workers from "./workers";
 import Logger from "./utils/logger";
+import * as EmailClient from "./init/email-client";
+import emailQueue from "./queues/email-queue";
```
Comment (Bruception) [inline backend/src/server.ts:18]: Along with this import

File: backend/src/server.ts (hunk 22 above)
Comment (Miodec) [reply backend/src/server.ts:85]: yes

File: backend/src/server.ts (hunk 23 above)
Comment (Miodec) [reply backend/src/server.ts:18]: yeap

File: backend/src/workers/email-worker.ts [hunk 24]
```diff
@@ -0,0 +1,35 @@
...
+  const result = await sendMailUsingTemplate(type, email, ctx);
+
+  if (!result.success) {
+    throw new Error(result.message);
+  }
+
+  const elapsed = performance.now() - start;
+
+  Logger.success(`Job: ${type} - completed in ${elapsed}ms`);
+}
```
Comment (Bruception) [inline backend/src/workers/email-worker.ts:29]: Squeaky clean

File: backend/src/init/email-client.ts [hunk 25]
```diff
@@ -0,0 +1,149 @@
...
+      message: "Email client transport not initialized",
+    };
+  }
+
+  const template = await fillTemplate<typeof templateName>(templateName, data);
+
+  const mailOptions = {
+    from: "Monkeytype <noreply@monkeytype.com>",
+    to,
+    subject: templates[templateName].subject,
```
Comment (Bruception) [inline backend/src/init/email-client.ts:109]: Should we add a guard to block even attempting to send an e-mail if the the templateName is invalid?

**Comment by monkeytypegeorge** [comment]
Continuous integration check(s) failed. Please review the failing check's logs and make the necessary changes. https://github.com/monkeytypegame/monkeytype/actions/runs/4059661758

**Comment by monkeytypegeorge** [comment]
Continuous integration check(s) failed. Please review the failing check's logs and make the necessary changes. https://github.com/monkeytypegame/monkeytype/actions/runs/4059839615

File: backend/src/init/email-client.ts (hunk 25 above)
Comment (Miodec) [reply backend/src/init/email-client.ts:100]: sure

Comment (Miodec) [reply backend/src/init/email-client.ts:100]: Well, actually - type script will block the function call if the parameter is not an `EmailType` and the `templates` object requires all of the types to be keys, so not sure if its needed.

File: backend/src/api/routes/users.ts [hunk 26]
```diff
@@ -582,4 +582,11 @@ router.post(
   asyncHandler(UserController.reportUser)
 );
 
+router.get(
+  "/requestVerificationEmail",
```
Comment (Bruception) [inline backend/src/api/routes/users.ts:586]: Just: “verificationEmail”, request is already implied

File: backend/src/api/controllers/user.ts [hunk 27]
```diff
@@ -58,6 +59,24 @@ export async function createNewUser(
   return new MonkeyResponse("User created");
 }
 
+export async function sendVerificationEmail(
+  req: MonkeyTypes.Request
+): Promise<MonkeyResponse> {
+  const { email, uid } = req.ctx.decodedToken;
+
+  const userInfo = await UserDAL.getUser(uid, "request verification email");
```
Comment (Bruception) [inline backend/src/api/controllers/user.ts:67]: Is there any way for us to exit early if the user has already been verified? Relying on client side UI is not good practice.

**Review by Bruception — COMMENTED** [review]
Lgtm, just one last question

**Pull request merged**